File: src/main/java/com/github/highright1234/glacier/protocol/Protocol.java
package com.github.highright1234.glacier.protocol;

import lombok.Getter;

import java.util.*;

@Getter
public class Protocol {

    public static Type HANDSHAKE = new Type();
    public static Type STATUS = new Type();
    public static Type LOGIN = new Type();
    public static Type PLAY = new Type();

    public static Set<Integer> SUPPORT_PROTOCOLS = new TreeSet<>();

    public static class Type {
        public DirectionData TO_CLIENT = new DirectionData();
        public DirectionData TO_SERVER = new DirectionData();

        public static class DirectionData {

            private final Map<Class<? extends AbstractPacket>, Map<Integer, Integer>> packetIdMap = new HashMap<>();
            private final Map<Integer, Map<Integer, AbstractPacket>> packetDataMap = new HashMap<>();

            public void addPacket(AbstractPacket packet, Map<Integer, Integer> mappingPacket) {
                packetIdMap.put(packet.getClass(), mappingPacket);
                for (int protocolVersion : mappingPacket.keySet()) {
                    packetDataMap.computeIfAbsent(protocolVersion, k -> new HashMap<>());
                    packetDataMap.get(protocolVersion).put(mappingPacket.get(protocolVersion), packet);
                }
            }

            public Integer getId(int protocolVersion, Class<? extends AbstractPacket> clazz) {
                return packetIdMap.get(clazz).get(protocolVersion);
            }

            public AbstractPacket getPacket(int protocolVersion, int id) {
                return packetDataMap.get(protocolVersion).get(id);
            }
        }
    }

    public static class Version {
        public static final int MINECRAFT_1_7_5 = 4;
        public static final int MINECRAFT_1_7_10 = 5;
        public static final int MINECRAFT_1_8 = 47;
        public static final int MINECRAFT_1_9 = 107;
        public static final int MINECRAFT_1_9_1 = 108;
        public static final int MINECRAFT_1_9_2 = 109;
        public static final int MINECRAFT_1_9_4 = 110;
        public static final int MINECRAFT_1_10 = 210;
        public static final int MINECRAFT_1_11 = 315;
        public static final int MINECRAFT_1_11_1 = 316;
        public static final int MINECRAFT_1_12 = 335;
        public static final int MINECRAFT_1_12_1 = 338;
        public static final int MINECRAFT_1_12_2 = 340;
        public static final int MINECRAFT_1_13 = 393;
        public static final int MINECRAFT_1_13_1 = 401;
        public static final int MINECRAFT_1_13_2 = 404;
        public static final int MINECRAFT_1_14 = 477;
        public static final int MINECRAFT_1_14_1 = 480;
        public static final int MINECRAFT_1_14_2 = 485;
        public static final int MINECRAFT_1_14_3 = 490;
        public static final int MINECRAFT_1_14_4 = 498;
        public static final int MINECRAFT_1_15 = 573;
        public static final int MINECRAFT_1_15_1 = 575;
        public static final int MINECRAFT_1_15_2 = 578;
        public static final int MINECRAFT_1_16 = 735;
        public static final int MINECRAFT_1_16_1 = 736;
        public static final int MINECRAFT_1_16_2 = 751;
        public static final int MINECRAFT_1_16_3 = 753;
        public static final int MINECRAFT_1_16_4 = 754;
        public static final int MINECRAFT_1_17 = 755;
        public static final int MINECRAFT_1_17_1 = 756;
    }
}

File: src/main/kotlin/com/github/highright1234/glacier/protocol/Protocol.kt
package com.github.highright1234.glacier.protocol

import java.util.HashMap
import java.lang.Error
import com.github.highright1234.glacier.protocol.packet.handshake.client.HandshakePacket
import com.github.highright1234.glacier.protocol.packet.status.client.SLPRequest
import com.github.highright1234.glacier.protocol.packet.status.client.Ping
import com.github.highright1234.glacier.protocol.packet.status.server.SLPResponse
import com.github.highright1234.glacier.protocol.packet.status.server.Pong
import com.github.highright1234.glacier.protocol.packet.login.client.LoginStart
import com.github.highright1234.glacier.protocol.packet.login.server.LoginSuccess
import com.github.highright1234.glacier.protocol.packet.DisconnectPacket
import com.github.highright1234.glacier.protocol.packet.login.server.EncryptionRequestPacket
import com.github.highright1234.glacier.protocol.packet.login.client.EncryptionResponsePacket
import com.github.highright1234.glacier.protocol.packet.play.PluginMessagePacket

class Protocol {

    val HANDSHAKE = Type()
    val STATUS = Type()
    val LOGIN = Type()
    val PLAY = Type()

    class Type {
        @JvmField
        var TO_CLIENT = DirectionData()
        @JvmField
        var TO_SERVER = DirectionData()

        class DirectionData {
            private val packetIdMap: MutableMap<Class<out MinecraftPacket>, Int?> = HashMap()
            private val packetDataMap: MutableMap<Int?, MinecraftPacket> = HashMap()
            fun addPacket(packet: MinecraftPacket, id: Int) {
                if (packetIdMap[packet.javaClass] != null) {
                    packetDataMap.remove(packetIdMap[packet.javaClass])
                }
                packetIdMap[packet.javaClass] = id
                packetDataMap[id] = packet
            }

            fun getId(clazz: Class<out MinecraftPacket>): Int? {
                return packetIdMap[clazz]
            }

            fun getPacket(id: Int): MinecraftPacket? {
                return packetDataMap[id]
            }

            fun getPacket(clazz: Class<out MinecraftPacket>): MinecraftPacket? {
                return packetDataMap[getId(clazz)]
            }
        }
    }

    object Version {
        // It's same protocol version of 1.7 - 1.7.5
        const val MINECRAFT_1_7_5 = 4

        // It's same protocol version of 1.7.6 - 1.7.10
        const val MINECRAFT_1_7_10 = 5

        // It's same protocol version of 1.8 - 1.8.9
        const val MINECRAFT_1_8_9 = 47
        const val MINECRAFT_1_9 = 107
        const val MINECRAFT_1_9_1 = 108
        const val MINECRAFT_1_9_2 = 109

        // It's same protocol version of 1.9.3 - 1.9.4
        const val MINECRAFT_1_9_4 = 110

        // It's same protocol version of 1.10 - 1.10.2
        const val MINECRAFT_1_10 = 210
        const val MINECRAFT_1_11 = 315
        const val MINECRAFT_1_11_1 = 316
        const val MINECRAFT_1_12 = 335
        const val MINECRAFT_1_12_1 = 338
        const val MINECRAFT_1_12_2 = 340
        const val MINECRAFT_1_13 = 393
        const val MINECRAFT_1_13_1 = 401
        const val MINECRAFT_1_13_2 = 404
        const val MINECRAFT_1_14 = 477
        const val MINECRAFT_1_14_1 = 480
        const val MINECRAFT_1_14_2 = 485
        const val MINECRAFT_1_14_3 = 490
        const val MINECRAFT_1_14_4 = 498
        const val MINECRAFT_1_15 = 573
        const val MINECRAFT_1_15_1 = 575
        const val MINECRAFT_1_15_2 = 578
        const val MINECRAFT_1_16 = 735
        const val MINECRAFT_1_16_1 = 736
        const val MINECRAFT_1_16_2 = 751
        const val MINECRAFT_1_16_3 = 753
        const val MINECRAFT_1_16_4 = 754
        const val MINECRAFT_1_17 = 755
        const val MINECRAFT_1_17_1 = 756
    }

    companion object {
        fun createProtocol(version: Int): Protocol {
            if (version < Version.MINECRAFT_1_7_5 || version > Version.MINECRAFT_1_17_1) {
                throw Error("version is unsupported version")
            }
            val out = Protocol()
            out.HANDSHAKE.TO_SERVER.addPacket(HandshakePacket(), 0x00)
            out.STATUS.TO_SERVER.addPacket(SLPRequest(), 0x00)
            out.STATUS.TO_SERVER.addPacket(Ping(), 0x00)
            out.STATUS.TO_CLIENT.addPacket(SLPResponse(""), 0x00)
            out.STATUS.TO_CLIENT.addPacket(Pong(), 0x00)
            out.LOGIN.TO_SERVER.addPacket(LoginStart(), 0x00)
            out.LOGIN.TO_CLIENT.addPacket(LoginSuccess(), 0x02)
            out.LOGIN.TO_CLIENT.addPacket(DisconnectPacket(), 0x00)
            out.LOGIN.TO_SERVER.addPacket(EncryptionRequestPacket(), 0x01)
            out.LOGIN.TO_CLIENT.addPacket(EncryptionResponsePacket(), 0x01)
            out.PLAY.TO_CLIENT.addPacket(PluginMessagePacket(), 0x3F)
            out.PLAY.TO_SERVER.addPacket(PluginMessagePacket(), 0x17)
            if (version >= Version.MINECRAFT_1_7_10) {
                if (version >= Version.MINECRAFT_1_8_9) {
                    if (version >= Version.MINECRAFT_1_9) {
                        if (version >= Version.MINECRAFT_1_9_1) {
                            if (version >= Version.MINECRAFT_1_9_2) {
                                if (version >= Version.MINECRAFT_1_9_4) {
                                    if (version >= Version.MINECRAFT_1_10) {
                                    }
                                }
                            }
                        }
                    }
                }
            }
            return out
        }
    }
}